diff --git a/src/evdev-mt-touchpad-gestures.c b/src/evdev-mt-touchpad-gestures.c
index 37682fb4..87ad0d0e 100644
--- a/src/evdev-mt-touchpad-gestures.c
+++ b/src/evdev-mt-touchpad-gestures.c
@@ -31,6 +31,7 @@
 
 #define DEFAULT_GESTURE_SWITCH_TIMEOUT ms2us(100)
 #define DEFAULT_GESTURE_2FG_SCROLL_TIMEOUT ms2us(150)
+#define DEFAULT_GESTURE_TAP_TIMEOUT ms2us(200)
 
 static inline const char*
 gesture_state_to_str(enum tp_gesture_state state)
@@ -41,6 +42,7 @@ gesture_state_to_str(enum tp_gesture_state state)
 	CASE_RETURN_STRING(GESTURE_STATE_SCROLL);
 	CASE_RETURN_STRING(GESTURE_STATE_PINCH);
 	CASE_RETURN_STRING(GESTURE_STATE_SWIPE);
+	CASE_RETURN_STRING(GESTURE_STATE_TAP);
 	}
 	return NULL;
 }
@@ -121,6 +123,11 @@ tp_gesture_start(struct tp_dispatch *tp, uint64_t time)
 				     tp->gesture.finger_count,
 				     &zero, &zero);
 		break;
+        case GESTURE_STATE_TAP:
+                gesture_notify_tap(&tp->device->base, time,
+                                     LIBINPUT_EVENT_GESTURE_TAP_BEGIN,
+                                   tp->gesture.finger_count);
+                break;
 	}
 
 	tp->gesture.started = true;
@@ -358,8 +365,19 @@ tp_gesture_handle_state_unknown(struct tp_dispatch *tp, uint64_t time)
 	/* Else wait for both fingers to have moved */
 	dir1 = tp_gesture_get_direction(tp, first, tp->gesture.finger_count);
 	dir2 = tp_gesture_get_direction(tp, second, tp->gesture.finger_count);
-	if (dir1 == UNDEFINED_DIRECTION || dir2 == UNDEFINED_DIRECTION)
-		return GESTURE_STATE_UNKNOWN;
+	if (dir1 == UNDEFINED_DIRECTION || dir2 == UNDEFINED_DIRECTION) {
+                struct device_float_coords raw;
+                struct normalized_coords delta, unaccel;
+                raw = tp_get_average_touches_delta(tp);
+                delta = tp_filter_motion(tp, &raw, time);
+                unaccel = tp_normalize_delta(tp, raw);
+                if (normalized_is_zero(delta) && normalized_is_zero(unaccel) &&
+                    time < (tp->gesture.initial_time + DEFAULT_GESTURE_TAP_TIMEOUT)) {
+                    // TODO: check tap whether enabled
+                    return GESTURE_STATE_TAP;
+                }
+                return GESTURE_STATE_UNKNOWN;
+        }
 
 	/* If both touches are moving in the same direction assume
 	 * scroll or swipe */
@@ -427,6 +445,54 @@ tp_gesture_handle_state_swipe(struct tp_dispatch *tp, uint64_t time)
 }
 
 static enum tp_gesture_state
+tp_gesture_handle_state_tap(struct tp_dispatch *tp, uint64_t time)
+{
+	struct device_float_coords raw;
+	struct normalized_coords delta, unaccel;
+	struct tp_touch *first = tp->gesture.touches[0],
+			*second = tp->gesture.touches[1];
+	int dir1, dir2;
+
+	if (time > (tp->gesture.initial_time + DEFAULT_GESTURE_TAP_TIMEOUT)) {
+		goto out;
+	}
+
+	raw = tp_get_average_touches_delta(tp);
+	delta = tp_filter_motion(tp, &raw, time);
+	unaccel = tp_normalize_delta(tp, raw);
+	if (normalized_is_zero(delta) && normalized_is_zero(unaccel)) {
+		tp_gesture_start(tp, time);
+		gesture_notify_tap(&tp->device->base, time,
+				LIBINPUT_EVENT_GESTURE_TAP_UPDATE,
+				tp->gesture.finger_count);
+		return GESTURE_STATE_TAP;
+	}
+
+out:
+	// TODO: optimize structure
+
+	/* Else wait for both fingers to have moved */
+	dir1 = tp_gesture_get_direction(tp, first, tp->gesture.finger_count);
+	dir2 = tp_gesture_get_direction(tp, second, tp->gesture.finger_count);
+
+	/* If both touches are moving in the same direction assume
+	 * scroll or swipe */
+	if (tp_gesture_same_directions(dir1, dir2)) {
+		if (tp->gesture.finger_count == 2) {
+			tp_gesture_set_scroll_buildup(tp);
+			return GESTURE_STATE_SCROLL;
+		} else if (tp->gesture.enabled) {
+			return GESTURE_STATE_SWIPE;
+		}
+	} else {
+		tp_gesture_init_pinch(tp);
+		return GESTURE_STATE_PINCH;
+	}
+
+	return GESTURE_STATE_UNKNOWN;
+}
+
+static enum tp_gesture_state
 tp_gesture_handle_state_pinch(struct tp_dispatch *tp, uint64_t time)
 {
 	double angle, angle_delta, distance, scale;
@@ -490,6 +556,10 @@ tp_gesture_post_gesture(struct tp_dispatch *tp, uint64_t time)
 		tp->gesture.state =
 			tp_gesture_handle_state_pinch(tp, time);
 
+	if (tp->gesture.state == GESTURE_STATE_TAP)
+            tp->gesture.state =
+                tp_gesture_handle_state_tap(tp, time);
+
 	evdev_log_debug(tp->device,
 			"gesture state: %s → %s\n",
 			gesture_state_to_str(oldstate),
@@ -569,6 +639,12 @@ tp_gesture_end(struct tp_dispatch *tp, uint64_t time, bool cancelled)
 					 tp->gesture.finger_count,
 					 cancelled);
 		break;
+	case GESTURE_STATE_TAP:
+            gesture_notify_tap_end(&tp->device->base,
+                                   time,
+                                   tp->gesture.finger_count,
+                                   cancelled);
+            break;
 	}
 
 	tp->gesture.started = false;
diff --git a/src/evdev-mt-touchpad.h b/src/evdev-mt-touchpad.h
index d24f77ee..9c95fd7e 100644
--- a/src/evdev-mt-touchpad.h
+++ b/src/evdev-mt-touchpad.h
@@ -134,6 +134,7 @@ enum tp_gesture_state {
 	GESTURE_STATE_SCROLL,
 	GESTURE_STATE_PINCH,
 	GESTURE_STATE_SWIPE,
+	GESTURE_STATE_TAP,
 };
 
 enum tp_thumb_state {
diff --git a/src/libinput-private.h b/src/libinput-private.h
index d50154ef..6a5847e6 100644
--- a/src/libinput-private.h
+++ b/src/libinput-private.h
@@ -578,6 +578,18 @@ gesture_notify_pinch_end(struct libinput_device *device,
 			 int cancelled);
 
 void
+gesture_notify_tap(struct libinput_device *device,
+                   uint64_t time,
+                   enum libinput_event_type type,
+                   int finger_count);
+
+void
+gesture_notify_tap_end(struct libinput_device *device,
+                       uint64_t time,
+                       int finger_count,
+                       int cancelled);
+
+void
 tablet_notify_axis(struct libinput_device *device,
 		   uint64_t time,
 		   struct libinput_tablet_tool *tool,
diff --git a/src/libinput.c b/src/libinput.c
index d85693a4..a6990b88 100644
--- a/src/libinput.c
+++ b/src/libinput.c
@@ -135,6 +135,9 @@ event_type_to_str(enum libinput_event_type type)
 	CASE_RETURN_STRING(LIBINPUT_EVENT_GESTURE_PINCH_BEGIN);
 	CASE_RETURN_STRING(LIBINPUT_EVENT_GESTURE_PINCH_UPDATE);
 	CASE_RETURN_STRING(LIBINPUT_EVENT_GESTURE_PINCH_END);
+	CASE_RETURN_STRING(LIBINPUT_EVENT_GESTURE_TAP_BEGIN);
+	CASE_RETURN_STRING(LIBINPUT_EVENT_GESTURE_TAP_UPDATE);
+	CASE_RETURN_STRING(LIBINPUT_EVENT_GESTURE_TAP_END);
 	CASE_RETURN_STRING(LIBINPUT_EVENT_SWITCH_TOGGLE);
 	case LIBINPUT_EVENT_NONE:
 		abort();
@@ -396,6 +399,9 @@ libinput_event_get_gesture_event(struct libinput_event *event)
 			   LIBINPUT_EVENT_GESTURE_SWIPE_BEGIN,
 			   LIBINPUT_EVENT_GESTURE_SWIPE_UPDATE,
 			   LIBINPUT_EVENT_GESTURE_SWIPE_END,
+			   LIBINPUT_EVENT_GESTURE_TAP_BEGIN,
+			   LIBINPUT_EVENT_GESTURE_TAP_UPDATE,
+			   LIBINPUT_EVENT_GESTURE_TAP_END,
 			   LIBINPUT_EVENT_GESTURE_PINCH_BEGIN,
 			   LIBINPUT_EVENT_GESTURE_PINCH_UPDATE,
 			   LIBINPUT_EVENT_GESTURE_PINCH_END);
@@ -879,6 +885,9 @@ libinput_event_gesture_get_time(struct libinput_event_gesture *event)
 	require_event_type(libinput_event_get_context(&event->base),
 			   event->base.type,
 			   0,
+			   LIBINPUT_EVENT_GESTURE_TAP_BEGIN,
+			   LIBINPUT_EVENT_GESTURE_TAP_UPDATE,
+			   LIBINPUT_EVENT_GESTURE_TAP_END,
 			   LIBINPUT_EVENT_GESTURE_PINCH_BEGIN,
 			   LIBINPUT_EVENT_GESTURE_PINCH_UPDATE,
 			   LIBINPUT_EVENT_GESTURE_PINCH_END,
@@ -895,6 +904,9 @@ libinput_event_gesture_get_time_usec(struct libinput_event_gesture *event)
 	require_event_type(libinput_event_get_context(&event->base),
 			   event->base.type,
 			   0,
+			   LIBINPUT_EVENT_GESTURE_TAP_BEGIN,
+			   LIBINPUT_EVENT_GESTURE_TAP_UPDATE,
+			   LIBINPUT_EVENT_GESTURE_TAP_END,
 			   LIBINPUT_EVENT_GESTURE_PINCH_BEGIN,
 			   LIBINPUT_EVENT_GESTURE_PINCH_UPDATE,
 			   LIBINPUT_EVENT_GESTURE_PINCH_END,
@@ -911,6 +923,9 @@ libinput_event_gesture_get_finger_count(struct libinput_event_gesture *event)
 	require_event_type(libinput_event_get_context(&event->base),
 			   event->base.type,
 			   0,
+			   LIBINPUT_EVENT_GESTURE_TAP_BEGIN,
+			   LIBINPUT_EVENT_GESTURE_TAP_UPDATE,
+			   LIBINPUT_EVENT_GESTURE_TAP_END,
 			   LIBINPUT_EVENT_GESTURE_PINCH_BEGIN,
 			   LIBINPUT_EVENT_GESTURE_PINCH_UPDATE,
 			   LIBINPUT_EVENT_GESTURE_PINCH_END,
@@ -927,6 +942,7 @@ libinput_event_gesture_get_cancelled(struct libinput_event_gesture *event)
 	require_event_type(libinput_event_get_context(&event->base),
 			   event->base.type,
 			   0,
+			   LIBINPUT_EVENT_GESTURE_TAP_END,
 			   LIBINPUT_EVENT_GESTURE_PINCH_END,
 			   LIBINPUT_EVENT_GESTURE_SWIPE_END);
 
@@ -939,6 +955,9 @@ libinput_event_gesture_get_dx(struct libinput_event_gesture *event)
 	require_event_type(libinput_event_get_context(&event->base),
 			   event->base.type,
 			   0.0,
+			   LIBINPUT_EVENT_GESTURE_TAP_BEGIN,
+			   LIBINPUT_EVENT_GESTURE_TAP_UPDATE,
+			   LIBINPUT_EVENT_GESTURE_TAP_END,
 			   LIBINPUT_EVENT_GESTURE_PINCH_BEGIN,
 			   LIBINPUT_EVENT_GESTURE_PINCH_UPDATE,
 			   LIBINPUT_EVENT_GESTURE_PINCH_END,
@@ -955,6 +974,9 @@ libinput_event_gesture_get_dy(struct libinput_event_gesture *event)
 	require_event_type(libinput_event_get_context(&event->base),
 			   event->base.type,
 			   0.0,
+			   LIBINPUT_EVENT_GESTURE_TAP_BEGIN,
+			   LIBINPUT_EVENT_GESTURE_TAP_UPDATE,
+			   LIBINPUT_EVENT_GESTURE_TAP_END,
 			   LIBINPUT_EVENT_GESTURE_PINCH_BEGIN,
 			   LIBINPUT_EVENT_GESTURE_PINCH_UPDATE,
 			   LIBINPUT_EVENT_GESTURE_PINCH_END,
@@ -972,6 +994,9 @@ libinput_event_gesture_get_dx_unaccelerated(
 	require_event_type(libinput_event_get_context(&event->base),
 			   event->base.type,
 			   0.0,
+			   LIBINPUT_EVENT_GESTURE_TAP_BEGIN,
+			   LIBINPUT_EVENT_GESTURE_TAP_UPDATE,
+			   LIBINPUT_EVENT_GESTURE_TAP_END,
 			   LIBINPUT_EVENT_GESTURE_PINCH_BEGIN,
 			   LIBINPUT_EVENT_GESTURE_PINCH_UPDATE,
 			   LIBINPUT_EVENT_GESTURE_PINCH_END,
@@ -989,6 +1014,9 @@ libinput_event_gesture_get_dy_unaccelerated(
 	require_event_type(libinput_event_get_context(&event->base),
 			   event->base.type,
 			   0.0,
+			   LIBINPUT_EVENT_GESTURE_TAP_BEGIN,
+			   LIBINPUT_EVENT_GESTURE_TAP_UPDATE,
+			   LIBINPUT_EVENT_GESTURE_TAP_END,
 			   LIBINPUT_EVENT_GESTURE_PINCH_BEGIN,
 			   LIBINPUT_EVENT_GESTURE_PINCH_UPDATE,
 			   LIBINPUT_EVENT_GESTURE_PINCH_END,
@@ -2696,6 +2724,29 @@ gesture_notify_pinch_end(struct libinput_device *device,
 }
 
 void
+gesture_notify_tap(struct libinput_device *device,
+                   uint64_t time,
+                   enum libinput_event_type type,
+                   int finger_count)
+{
+        const struct normalized_coords zero = { 0.0, 0.0 };
+        gesture_notify(device, time, type, finger_count, 0, &zero, &zero,
+                       0.0, 0.0);
+}
+
+void
+gesture_notify_tap_end(struct libinput_device *device,
+                       uint64_t time,
+                       int finger_count,
+                       int cancelled)
+{
+        const struct normalized_coords zero = { 0.0, 0.0 };
+
+        gesture_notify(device, time, LIBINPUT_EVENT_GESTURE_TAP_END,
+                       finger_count, cancelled, &zero, &zero, 0.0, 0.0);
+}
+
+void
 switch_notify_toggle(struct libinput_device *device,
 		     uint64_t time,
 		     enum libinput_switch sw,
@@ -3138,6 +3189,9 @@ libinput_event_gesture_get_base_event(struct libinput_event_gesture *event)
 			   LIBINPUT_EVENT_GESTURE_SWIPE_BEGIN,
 			   LIBINPUT_EVENT_GESTURE_SWIPE_UPDATE,
 			   LIBINPUT_EVENT_GESTURE_SWIPE_END,
+			   LIBINPUT_EVENT_GESTURE_TAP_BEGIN,
+			   LIBINPUT_EVENT_GESTURE_TAP_UPDATE,
+			   LIBINPUT_EVENT_GESTURE_TAP_END,
 			   LIBINPUT_EVENT_GESTURE_PINCH_BEGIN,
 			   LIBINPUT_EVENT_GESTURE_PINCH_UPDATE,
 			   LIBINPUT_EVENT_GESTURE_PINCH_END);
diff --git a/src/libinput.h b/src/libinput.h
index f1a0a2a6..ec063feb 100644
--- a/src/libinput.h
+++ b/src/libinput.h
@@ -806,6 +806,9 @@ enum libinput_event_type {
 	LIBINPUT_EVENT_GESTURE_PINCH_BEGIN,
 	LIBINPUT_EVENT_GESTURE_PINCH_UPDATE,
 	LIBINPUT_EVENT_GESTURE_PINCH_END,
+	LIBINPUT_EVENT_GESTURE_TAP_BEGIN,
+	LIBINPUT_EVENT_GESTURE_TAP_UPDATE,
+	LIBINPUT_EVENT_GESTURE_TAP_END,
 
 	LIBINPUT_EVENT_SWITCH_TOGGLE = 900,
 };
diff --git a/test/litest.c b/test/litest.c
index a31e874e..d64ecf59 100644
--- a/test/litest.c
+++ b/test/litest.c
@@ -2481,6 +2481,15 @@ litest_event_type_str(enum libinput_event_type type)
 	case LIBINPUT_EVENT_GESTURE_PINCH_END:
 		str = "GESTURE PINCH END";
 		break;
+	case LIBINPUT_EVENT_GESTURE_TAP_BEGIN:
+            str = "GESTURE TAP START";
+            break;
+	case LIBINPUT_EVENT_GESTURE_TAP_UPDATE:
+            str = "GESTURE TAP UPDATE";
+            break;
+	case LIBINPUT_EVENT_GESTURE_TAP_END:
+            str = "GESTURE TAP END";
+            break;
 	case LIBINPUT_EVENT_TABLET_TOOL_AXIS:
 		str = "TABLET TOOL AXIS";
 		break;
diff --git a/tools/libinput-debug-events.c b/tools/libinput-debug-events.c
index 33f51d0e..3829be54 100644
--- a/tools/libinput-debug-events.c
+++ b/tools/libinput-debug-events.c
@@ -119,6 +119,15 @@ print_event_header(struct libinput_event *ev)
 	case LIBINPUT_EVENT_GESTURE_PINCH_END:
 		type = "GESTURE_PINCH_END";
 		break;
+	case LIBINPUT_EVENT_GESTURE_TAP_BEGIN:
+            type = "GESTURE_TAP_BEGIN";
+            break;
+	case LIBINPUT_EVENT_GESTURE_TAP_UPDATE:
+            type = "GESTURE_TAP_UPDATE";
+            break;
+	case LIBINPUT_EVENT_GESTURE_TAP_END:
+            type = "GESTURE_TAP_END";
+            break;
 	case LIBINPUT_EVENT_TABLET_TOOL_AXIS:
 		type = "TABLET_TOOL_AXIS";
 		break;
@@ -625,6 +634,7 @@ print_gesture_event_without_coords(struct libinput_event *ev)
 	type = libinput_event_get_type(ev);
 
 	if (type == LIBINPUT_EVENT_GESTURE_SWIPE_END ||
+	    type == LIBINPUT_EVENT_GESTURE_TAP_END ||
 	    type == LIBINPUT_EVENT_GESTURE_PINCH_END)
 	    cancelled = libinput_event_gesture_get_cancelled(t);
 
@@ -827,6 +837,15 @@ handle_and_print_events(struct libinput *li)
 		case LIBINPUT_EVENT_GESTURE_PINCH_END:
 			print_gesture_event_without_coords(ev);
 			break;
+		case LIBINPUT_EVENT_GESTURE_TAP_BEGIN:
+                    print_gesture_event_without_coords(ev);
+                    break;
+		case LIBINPUT_EVENT_GESTURE_TAP_UPDATE:
+                    print_gesture_event_with_coords(ev);
+                    break;
+		case LIBINPUT_EVENT_GESTURE_TAP_END:
+                    print_gesture_event_without_coords(ev);
+                    break;
 		case LIBINPUT_EVENT_TABLET_TOOL_AXIS:
 			print_tablet_axis_event(ev);
 			break;
diff --git a/tools/libinput-debug-gui.c b/tools/libinput-debug-gui.c
index ab49ad66..3b079d1e 100644
--- a/tools/libinput-debug-gui.c
+++ b/tools/libinput-debug-gui.c
@@ -90,6 +90,11 @@ struct window {
 		double x, y;
 	} pinch;
 
+        struct {
+            int nfingers;
+            double x, y;
+        } tap;
+
 	struct {
 		double x, y;
 		double x_in, y_in;
@@ -832,6 +837,11 @@ handle_event_libinput(GIOChannel *source, GIOCondition condition, gpointer data)
 		case LIBINPUT_EVENT_GESTURE_PINCH_END:
 			handle_event_pinch(ev, w);
 			break;
+		case LIBINPUT_EVENT_GESTURE_TAP_BEGIN:
+		case LIBINPUT_EVENT_GESTURE_TAP_UPDATE:
+		case LIBINPUT_EVENT_GESTURE_TAP_END:
+                    // TODO: draw tap gesture
+                    break;
 		case LIBINPUT_EVENT_TABLET_TOOL_AXIS:
 		case LIBINPUT_EVENT_TABLET_TOOL_PROXIMITY:
 		case LIBINPUT_EVENT_TABLET_TOOL_TIP:
